Release 0.3.17 (template updates) - ??

joshbruce opened the pull request:
## Description

- Fixes #1075 

## Submitter

- [x] `$ npm version` has been run.
- [x] Release notes in [draft GitHub release](https://github.com/markedjs/marked/releases) are up to date
- [x] Reviewer checklist is complete.
- [ ] Merge PR.
- [ ] Publish GitHub release using `master` with correct version number.
- [ ] `$ npm publish` has been run.
- [ ] Create draft GitHub release to prepare next release.

Note: If merges to `master` occur after submitting this PR and before running `$ npm pubish` you should be able to

1. pull from `upstream/master` into the branch holding this version,
2. run `$ npm run build` to regenerate the `min` file, and
3. commit and push the updated changes.

## Reviewer

- [x] Version in `package.json` has been updated (see [RELEASE.md](https://github.com/markedjs/marked/blob/master/RELEASE.md)).
- [x] The `marked.min.js` has been updated; or,
- [ ] release does not change library.
- [x] case_insensitive_refs is only failing test (remove once CI is in place and all tests pass).
- [x] All lint checks pass (remove once CI is in place).

joshbruce commented:
@Feder1co5oave: Are we in the forcing merges despite failing tests phase??

Feder1co5oave commented:
Currently, we are. We haven't been in an "all tests passing" status for a long time. Once we fix that test case, we'll make sure that new PRs don't introduce new failing tests.

joshbruce commented:
Think this now covers all the feedback in #1075 except what's under the "input" header - maybe a different issue or conversation. Taking out the WIP part.

joshbruce commented:
Note: Leaving in the possibility that the release doesn't actually change the library because we may need to change the README or something without actually changing the code...not sure NPM updates that without publishing a version...could be totally wrong (NPM is definitely different than Packagist).

styfle commented:
NPM requires publishing a new version to update the README. 

It should be a patch update like with 

    npm version patch

joshbruce commented:
I suppose I could always rollback on my branch but let's see if we can do it this way.

1. In the previous sequence publishing to NPM happened before the PR; therefore, the previous #1067 (release 0.3.16) was already published to NPM but not merged before the lint and Travis setup...my bad and hopefully not something we have to worry about moving forward.
2. Went to fix the merge conflicts on the other PR but a) really can't update anything to that version (see previous); so, no GitHub release for 0.3.16 and b) not sure but it seems GitHub or I lost each other anyway. Therefore, this branch got the update from master.
3. Added @UziTech's recommendation to automatically commit the minification step...still feels weird to me, at least on this PR because I use GitHub Desktop but I'm cool with it if everyone else is as well.
4. Accepted the recommendation from @Feder1co5oave on order of operations - starting with version and then PR before doing anything else; therefore, this little hiccup shouldn't happen again.

@Feder1co5oave, @UziTech, and @styfle: Broke and fixed some things.

Feder1co5oave commented:
I'm not following what this is about. Is this a release PR?

joshbruce commented:
@Feder1co5oave: It became one yes. Didn't start as one, no. Not sure how to explain the order of operations that got us here, to be honest:

1. There was a fault in the process that got 0.3.16 to NPM - my bad.
2. So, when we merged in the Travis + lint PR before finishing out the 0.3.16 release PR - the 0.3.16 PR had to be closed.
3. Therefore, I updated the process ordering to account for that based on your feedback; or, at least my interpretation of it (having the GitHub release get done prior to `npm publish`).
4. When I updated this branch after the Travis + lint PR, it came with the updated `min`.

So, that's how this ended up becoming a release PR despite not starting that way. And how the process got updated to what you see in the checklist right now: GitHub goes first, which I do like better, because it should allow us to bring in PRs on the fly before committing to a release.

Apologies for my part in the bumpy ride.

Feder1co5oave commented:
Wait, so we actually released 0.3.16 on npm?! I'm kinda lost.

I want to be clear on something because I'm seeing a lot of confusion:
### Release PRs should not include changes by themselves! They're only for releasing on npm. So basically, what should change is only package.json with the new version, and probably marked.min.js to make sure it is updated with marked.js!

Here, you're trying to merge (several) changes which pertain to very different aspect of the development. None of which bring any code change or README change, so these changes have no reason to be released to npm! You're also trying to release.

joshbruce commented:
@Feder1co5oave: I'm not sure I'm catching the tone properly from the text.

1. We're changing processes and doing things a lot of us have not really done before (and never have done together); therefore, there are bound to be hiccups, if they don't improve, then we freak out. :)

2. Yes. The previous process I used for releases was to publish to NPM before doing the GitHub release. This was the first time in all the releases we have done since I took over publishing that something went pear-shaped. Therefore, the process got updated (improved) and we should have that problem moving forward - no matter who is doing the committing - newbie off the street or a veteran...because we have the checklist.

3. `npm publish` will put everything from the repo at the time the command is run and not in the `.npmignore` file into the package published to NPM. The only thing in `.npmignore` is the test folder. Therefore, everything else in `master` winds up as part of the NPM package. (And, honestly, I'm not sure the `.npmignore` is doing the proper thing yet - haven't got there.)

While I typically agree that a release will typically only have updates to the `package.json` and `min`, it doesn't have to be a hard and fast rule. If there's not a problem with the templates, then we can merge this, finish the checklists, and move on. If there's a serious issue with the template updates, then we might try something else.

joshbruce commented:
[edit]: Should not have that problem again...

We also have bigger fish to fry, I think right now regarding the REDOS thing. The changes being made by this PR, beyond the actual package code, are minor and can be easily changed in future.

Feder1co5oave commented:
I really thought it obvious that release PRs should only be about releasing and nothing else. Since it is not, I'm now proposing that they are, and to make this a hard rule to follow.

Releasing to npm just to update template changes that are specific to github is nonsense, because they're used only by github. We shouldn't release unless we've got a real reason to.

Is [this](https://github.com/markedjs/marked/releases/tag/v0.3.17) supposed to be a draft?

Feder1co5oave commented on package.json:
```diff
@@ -41,7 +41,7 @@
     "lint": "eslint --fix lib/marked.js test/index.js",
     "build": "uglifyjs lib/marked.js -cm  --comments /Copyright/ -o marked.min.js",
-    "preversion": "npm run build"
+    "preversion": "npm run build && (git diff --quiet || git commit -am 'minify')"
```
what's this?

joshbruce replied:
This was a recommendation from @UziTech. Basically instead of build then commit `min`...let NPM + Git do it for us. As far as the details of what Git is actually doing I would have to read the docs.

styfle commented:
> We shouldn't release unless we've got a real reason to

I agree with @Feder1co5oave 💯

joshbruce commented:
@styfle: I agree with @Feder1co5oave as well regarding releasing when we need to not *because* of some updates to documentation, which is not what caused this PR to become a release. And we have a reason. I really feel like I'm getting a little beat up over something that I was actually doing to try and help fix a problem. Not only that, but now we have a bunch of reasons to do a release.

@Feder1co5oave: Yes, that's supposed to be a draft - I suppose in my zeal and excitement of getting testing updated, the DOS solution in place, having solid docs up on the board, improving the processes and templates for us, having our first contributor who wasn't the four of us report and fix something, and so on, I accidentally hit publish instead of save draft...again, accidents happen...

Gotta step away for a minute, I think...been a busy day on here.

joshbruce commented:
Actually I still see it as a draft release...with all the updates as to why it's a release. And, once we merge in #1083 - we got one more reason to do a release.

<img width="788" alt="screen shot 2018-02-26 at 7 59 11 pm" src="https://user-images.githubusercontent.com/15252830/36704737-a6fcd962-1b2f-11e8-97e4-2d8f943d9a76.png">

styfle commented:
@joshbruce 
I think the problem we are getting at is that there should be two separate PRs:
1. Update GitHub issue templates
2. Bump version and release

To quote @Feder1co5oave 

> release PRs should only be about releasing and nothing else. Since it is not, I'm now proposing that they are, and to make this a hard rule to follow.

Feder1co5oave commented:
@joshbruce I now see it as a draft too. I think the new v0.3.17 tag confused me(?)

joshbruce commented:
@styfle: Agreed. I accidentally pulled master in a way that brought the release-worthy PR submission from @Feder1co5oave into this branch because I was on it.

Given the overall state of things - I don't think it's the end of the world. NPM and GitHub releases are two completely different things - and neither one has actually happened yet for 0.3.17.

So, if we review the templates - without getting hung up on how we got here - approve them (which I've been using them all day today anyway). 

1. We can merge in the DOS fix #1083. 
2. Possibly merge in the documentation fixes (to update README and all that on NPM as well, because as was pointed out by @styfle, the only way to do that is to publish a release to NPM). 
3. We can then update this branch again with an updated `min`.

Then our potential future contributors get updated docs (including the AUTHORS file to start working on governance and community things), we get update checklists that are cleaner, and even a security vulnerability check to boot.

@Feder1co5oave: Yeah. The tag thing confused me as well. Not sure what's going on there.

ps. Even if we don't like these templates...we can update them in pretty short order. I'm more concerned with Marked ending up on the blacklist again and us sticking so much to process that we can't handle accidents committed by contributors easily...including within the ring of committers.

joshbruce commented:
We also get the improvements to the testing harness submitted by @UziTech and @davisjam. Yes, they should be two separate PRs. Yes, up to this point, they always have been. Yes, in the future, given the updated checklists, it should be harder to have something like this happen again.
